langgraph/prebuilt/tool_validator.py

```diff
@@ -0,0 +1,225 @@
+
+    It can be used either in StateGraph with a "messages" key or in MessageGraph.
+
+    !!! note
+
+        This node does not actually **run** the tools, it only validates the tool calls,
+        which is useful for extraction and other use cases where you need to generate
+        structured output that conforms to a complex schema without losing the original
+        messages and tool IDs (for use in multi-turn conversations).
+
+    Args:
+        schemas: A list of schemas to validate the tool calls with. These can be
+            any of the following:
+            - A pydantic BaseModel class
+            - A BaseTool instance (the args_schema will be used)
+            - A function (a schema will be created from the function signature)
+        format_error: A function that takes an exception, a ToolCall, and a schema
+            and returns a formatted error string. By default, it returns the
+            exception repr and a message to respond after fixing validation errors.
+        name: The name of the node.
+        tags: A list of tags to add to the node.
+
+    Returns:
+        (Union[Dict[str, List[ToolMessage]], Sequence[ToolMessage]]): A list of ToolMessages with the validated content or error messages.
+
+    Examples:
+        Example usage for re-prompting the model to generate a valid response:
+        >>> from typing import Literal
+        ...
+        >>> from langchain_anthropic import ChatAnthropic
+        >>> from langchain_core.pydantic_v1 import BaseModel, validator
+        ...
+        >>> from langgraph.graph import END, START, MessageGraph
+        >>> from langgraph.prebuilt import ValidationNode
+        ...
+        ...
+        >>> class SelectNumber(BaseModel):
+        ...     a: int
+        ...
+        ...     @validator("a")
+        ...     def a_must_be_meaningful(cls, v):
+        ...         if v != 37:
+        ...             raise ValueError("Only 37 is allowed")
+        ...         return v
+        ...
+        ...
+        >>> builder = MessageGraph()
+        >>> llm = ChatAnthropic(model="claude-3-haiku-20240307").bind_tools([SelectNumber])
+        >>> builder.add_node("model", llm)
+        >>> builder.add_node("validation", ValidationNode([SelectNumber]))
+        >>> builder.add_edge(START, "model")
+        ...
+        ...
+        >>> def should_validate(state: list) -> Literal["validation", "__end__"]:
+        ...     if state[-1].tool_calls:
+        ...         return "validation"
+        ...     return END
+        ...
+        ...
+        >>> builder.add_conditional_edges("model", should_validate)
+        ...
+        ...
+        >>> def should_reprompt(state: list) -> Literal["model", "__end__"]:
+        ...     for msg in state[::-1]:
+        ...         # None of the tool calls were errors
+        ...         if msg.type == "ai":
+        ...             return END
+        ...         if msg.additional_kwargs.get("is_error"):
```

Comment:
A little sad we can't just return dict content in messages

Reply:
ya ... :(